riscv/csrs.cc

```diff
@@ -1240,3 +1240,110 @@ bool vxsat_csr_t::unlogged_write(const reg_t val) noexcept {
   dirty_vs_state;
   return masked_csr_t::unlogged_write(val);
 }
+
+// implement class stateen_csr_t
+stateen_csr_t::stateen_csr_t(processor_t* const proc, const reg_t addr, const reg_t mask,
+                             const reg_t init):
+  masked_csr_t(proc, addr, mask, init) {
+}
+
+void stateen_csr_t::verify_permissions(insn_t insn, bool write) const {
+  if (!proc->extension_enabled(EXT_SMSTATEEN))
+    throw trap_illegal_instruction(insn.bits());
+  masked_csr_t::verify_permissions(insn, write);
+}
+
+// implement class stateenh_csr_t
+stateenh_csr_t::stateenh_csr_t(processor_t* const proc, const reg_t addr, const reg_t mask,
+                               stateen_csr_t_p stateen):
+  csr_t(proc, addr),
+  stateen(stateen),
+  mask(mask) {
+}
+
+reg_t stateenh_csr_t::read() const noexcept {
+  return (stateen->read() >> 32) & mask;
+}
+
+bool stateenh_csr_t::unlogged_write(const reg_t val) noexcept {
+  return stateen->unlogged_write((stateen->written_value() & ~(mask << 32)) | ((val & mask) << 32));
+}
```

Comment:
This is unfortunate that each of the upper 32-bit *h regs have to have all this boilerplate. Can you find a way to modify & rename `mstatush_csr_t` so that you can use that class for each of these *h regs?

Reply:
OK.